### PR TITLE
GitHub connection error

### LINKED ISSUE #4668: Return an error if GitHub account is already linked to a Spectrum account
Thread: https://spectrum.chat/spectrum/general/cant-connect-my-github-account~298e1410-3ff6-4868-81d0-e6b8b0be4339

### PR DESCRIPTION
**Pull request by brianlovin**
**Status**
- [ ] WIP
- [x] Ready for review
- [ ] Needs testing

**Deploy after merge (delete what needn't be deployed)**
- api
- hyperion (frontend)

@mxstbr this feels...hacky. But I'm not quite sure how to else to approach this.

Here's what's happening:
1. User A logs in to Spectrum with GitHub
2. User A creates another Spectrum account accidentally (or intentionally) by logging in with Twitter
3. User A+Twitter tries to connect their GitHub account in their user settings
4. Our auth flow detects that an account already exists with those GitHub credentials
5. We return User A+Twitter back to their user settings, without editing their user object

The problem is people getting redirected back, seeing the 'Connect GitHub' button again, and assuming the app is broken.

This PR addresses this by passing back an error message as a query parameter. When the user settings view mounts, I check for the query param and use it to dispatch a toast and save a class variable with the error to be passed as props down to the GitHub connect button so that we can render this:

screenshot 2019-02-19 15 40 51

I'm open to ideas on this; the only obvious abstraction I can think of here would be to have a global query parser that looks for some generic `?error=foo` query param in every url in order to dispatch a toast. But I'm not sure if we want to do that yet, since we don't really have any other use cases for it, nor am I sure that it's the right way for us to handle error messages.

Closes #4668

### CODE REVIEW

**Review by mxstbr — COMMENTED** [review]
I would rather make this something generic like `error` from the get-go, we might need to reuse that for other auth or similar errors!

**Comment by brianlovin** [comment]
Fair enough - a good use case I just thought of would be to redirect all email unsubscribe links back to the app instead of that ugly unstyled html page. We could redirect it back with some kind of query params like `?toastType=success&toastText=Unsubscribe successful!` and then for this PRs use case it could be `?toastType=error&toastText=Blah...`

Then I'll just need an app-level history listener to handle locations with these params.

That seem better?

**Comment by mxstbr** [comment]
That seems great to me! :100:

**Comment by brianlovin** [comment]
Okay, I've implemented a generic toast handler component at the global level, and demonstrated a proof of concept implementation for email unsub redirects here: https://github.com/withspectrum/spectrum/pull/4679/commits/788143b9dcbf46247ac9a8765912807d2536331f => everything works that I've tested locally, but I'm going to write some E2E tests to make sure things look good!

**Comment by brianlovin** [comment]
Okay, this is ready for a test and review @mxstbr! I added some end to end tests as well to make sure this works :)

**Review by mxstbr — APPROVED** [review]
Amazing, this works really well!! Liking the new styling too, much cleaner ✨ 

Let's shipit!

File: src/views/queryParamToastDispatcher/index.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+      if (!key || key === 'toastMessage' || key === 'toastType') return null;
+      clean[key] = params[key];
+      return null;
+    });
+    const cleanParams = querystring.stringify(clean);
+    /*
+      We decode the cleanParams in order to preserver special characters in the url
+      For example, the url /spectrum/general/another-thread~thread-2?m=MTQ4MzIyNTIwMDAwMg== 
+      has two equals signs at the end. If we don't decode the cleanParams it will become
+      spectrum/general/another-thread~thread-2?m=MTQ4MzIyNTIwMDAwMg%3D%3D
```
Comment (mxstbr) [inline src/views/queryParamToastDispatcher/index.js:79]: Great catch here :100:

**Review by mxstbr — APPROVED** [review]
I am so shipping this!

**Comment by mxstbr** [comment]
Restarting e2e tests since failures seem unrelated.

Then I am shipping this.

😅

**Pull request merged**